cpp/src/arrow/dataset/file_base.cc

```diff
@@ -41,5 +45,68 @@ Status FileBasedDataFragment::Scan(std::shared_ptr<ScanContext> scan_context,
   return format_->ScanFile(source_, scan_options_, scan_context, out);
 }
 
+FileSystemBasedDataSource::FileSystemBasedDataSource(
+    fs::FileSystem* filesystem, const fs::Selector& selector,
+    std::shared_ptr<FileFormat> format, std::shared_ptr<ScanOptions> scan_options,
+    std::vector<fs::FileStats> stats)
+    : filesystem_(filesystem),
+      selector_(std::move(selector)),
+      format_(std::move(format)),
+      scan_options_(std::move(scan_options)),
+      stats_(std::move(stats)) {}
+
+Status FileSystemBasedDataSource::Make(fs::FileSystem* filesystem,
+                                       const fs::Selector& selector,
+                                       std::shared_ptr<FileFormat> format,
+                                       std::shared_ptr<ScanOptions> scan_options,
+                                       std::unique_ptr<FileSystemBasedDataSource>* out) {
+  std::vector<fs::FileStats> stats;
+  RETURN_NOT_OK(filesystem->GetTargetStats(selector, &stats));
+
+  auto new_end =
+      std::remove_if(stats.begin(), stats.end(), [&](const fs::FileStats& stats) {
```

Comment:
It seems this is potentially `O(n**2)` as each remove will shift all trailing vector elements to the left... Better build a new vector from scratch perhaps?

Reply:
remove_if is O(n) https://en.cppreference.com/w/cpp/algorithm/remove

Reply:
Ah, you're right, the signature allows that. Thanks.